Fix plugin loggers on shutdown

josemmo opened the pull request:
- Added stop signal listener to MinecraftServer
- Fixes #5270

> When system property "terminal.jline" is false and the server receives a SIGINT/SIGTERM plugin, non-root loggers did not work due to how shutdown hooks work.

josemmo committed:
Fix plugin loggers on shutdown (#5270)

- Added stop signal listener to MinecraftServer

> When system property "terminal.jline" is false and the server receives a SIGINT/SIGTERM plugin, non-root loggers did not work due to how shutdown hooks work.

Proximyst reviewed (commented):
```
warning: sun.misc.SignalHandler is Sun proprietary API and may be removed in a future release
```

Dunno if we want to rely on this...

HexedHero commented on Spigot-Server-Patches/0723-Fix-plugin-loggers-on-server-shutdown.patch:
```diff
@@ -0,0 +1,27 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: =?UTF-8?q?Jos=C3=A9=20Miguel=20Moreno?= <josemmo@pm.me>
```
The hell is this author-name

zml2008 replied:
https://en.wikipedia.org/wiki/MIME#Encoded-Word

josemmo committed:
Revert signal handlers patch

- Removed patch for #5270 that used signal handlers (4dd4295885dcb770666acb89d65962c14b101cbd)

josemmo committed:
Add custom log manager

- Created custom log manager to ignore calls to CustomLogManager#reset() before server has shutdown

> When system property "terminal.jline" is false and the server receives a SIGINT/SIGTERM plugin, non-root loggers did not work due to how shutdown hooks work.

Fixes #5270

Proximyst commented on Spigot-Server-Patches/0739-Fix-plugin-loggers-on-server-shutdown.patch:
```diff
@@ -0,0 +1,67 @@
+--- /dev/null
++++ b/src/main/java/com/destroystokyo/paper/log/CustomLogManager.java
+@@ -0,0 +1,26 @@
++package com.destroystokyo.paper.log;
```
`io.papermc.paper.log`

Proximyst reviewed (approved):
Clean! Nice!

josemmo committed:
Renamed CustomLogManager package

- Renamed com.destroystokyo.paper.log.CustomLogManager to io.papermc.paper.log.CustomLogManager
- Updated patch number

A248 commented on Spigot-Server-Patches/0755-Fix-plugin-loggers-on-server-shutdown.patch:
```diff
@@ -0,0 +1,75 @@
+ 
++    // Paper start - Hijack log manager to ensure logging on shutdown
++    static {
++        System.setProperty("java.util.logging.manager", io.papermc.paper.log.CustomLogManager.class.getName());
```
Are you sure `CustomLogManager.class` won't accidentally initialize LogManager? If you init LogManager too early, setting the property will happen too late.

I trust if you say that this worked during your testing.

DenWav replied:
I think this would be safest as string literal even if it does work - removes any possibly ambiguity.

josemmo replied:
> Are you sure `CustomLogManager.class` won't accidentally initialize LogManager?

The "java.util.logging.manager" property is being set as soon as the Main class is loaded and AFAIK there's no other place in the source code where this property gets overwritten. Tested it and it works, can't think of any reason why it shouldn't.

> I think this would be safest as string literal even if it does work - removes any possibly ambiguity.

The reason I don't use a string literal is because the app will crash if it fails to resolve the FQN, an error that otherwise will go unnoticed.

Proximyst replied:
If that's the case, I think using a string is fine. If it crashes when you rename it, you've obviously done something wrong. Put a comment in the class and I'm sure it's all good.

josemmo replied:
Ok, I'll change it. But bear in mind if someone refactors CustomLogManager their IDE will not recognize the string literal as a FQN and will not rename it.

A248 reviewed (commented):
More of a question than a comment.

josemmo committed:
Replace custom logger FQN with literal

- Used string literal when setting "java.util.logging.manager" property

Machine-Maker commented on Spigot-Server-Patches/0755-Fix-plugin-loggers-on-server-shutdown.patch:
```diff
@@ -0,0 +1,75 @@
+ import io.netty.buffer.Unpooled;
+ import io.papermc.paper.adventure.PaperAdventure; // Paper
+ import io.papermc.paper.event.entity.EntityMoveEvent;
++import io.papermc.paper.log.CustomLogManager;
```
Don't import here, use FQN below

josemmo replied:
I don't know whether that's possible. When using FQN instead of an import I get the following error trying to package the Paper-Server module:
```
Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.8.1:compile (default-compile) on project paper: Compilation failure
/home/josemmo/PaperMC/Paper-Server/src/main/java/net/minecraft/server/MinecraftServer.java:[973,11] cannot find symbol
  symbol:   variable papermc
  location: class net.minecraft.server.io
```

Besides, all other `io.papermc.*` classes used in `net.minecraft.server.MinecraftServer` have an import statement.

electronicboy replied:
Fully qualified in 1.16.5 won't work because of the ever loving `io` class, this will work fine in 1.17

josemmo replied:
Is there any documentation on how to build Paper 1.17? I'm trying to rebase this branch and can't even apply patches to master at commit 98fa13609982b3ee19928efed96f4d23735ea100:

```
> Task :patchCraftBukkitPatches FAILED
error: patch failed: .gradle/caches/paperweight/taskCache/patchCraftBukkitPatches.zip-1711287088/net/minecraft/world/entity/ai/behavior/BehaviorWorkComposter.patch:1
error: .gradle/caches/paperweight/taskCache/patchCraftBukkitPatches.zip-1711287088/net/minecraft/world/entity/ai/behavior/BehaviorWorkComposter.patch: patch does not apply        
FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':patchCraftBukkitPatches'.
> io.papermc.paperweight.PaperweightException: Command finished with 1 exit code: git -c commit.gpgsign=false -c core.safecrlf=false apply C:\Users\josemmo\Desktop\PaperMC\build-data\craftbukkit-patch-patches\BehaviorWorkComposter.patch.patch
```

The failing git command prints the following error message:
```
error: net/minecraft/world/entity/ai/behavior/BehaviorWorkComposter.patch: No such file or directory
```

electronicboy committed:
Merge branch 'master' into bugfix/shutdown-loggers

electronicboy committed:
Rebuild for 1.19
